### PR TITLE
246/auto shutdown on low battery voltage

### PR DESCRIPTION
**Pull request by malinradstrom**
Title:
246/auto-shutdown-on-low-battery-voltage

Description:
This pull request introduces automatic shutdown functionality when the battery voltage drops below a predefined threshold. Changes include:
- Arduino CommandHandler
- Added handling for the new JSON commands: BATTERY_LEVEL, LOW_BATTERY_WARNING, and SLEEP_MODE.
- On SLEEP_MODE: true, the Arduino sends an ACK and the Raspberry Pi will initiate a safe shutdown.
- Raspberry Pi Command Handler
- Extended process_command to forward incoming battery data in three parts:
- battery_level (float 0.00–1.00)
- low_battery_warning (boolean)
- sleep_mode (boolean)
- Added helper functions in battery.py to format and send each of these messages to the Arduino.
- Implemented incoming‐sleep handling that relays the shutdown command to the Pi and triggers sudo shutdown now when appropriate.

Unit Tests:
New and updated pytest suites verify correct JSON payloads for all three battery‐related commands, including edge values (0.00, 1.00), and confirm the shutdown trigger response.

All existing functionality (steering, lights, camera, streaming) remains unchanged. CI workflows now pass with 100% green across both Arduino and Python test suites.

### CODE REVIEW

File: Embedded/RaspberryPI/communication/serial_helper.py [hunk 1]
```diff
@@ -15,7 +15,7 @@ def __init__(self):
         self.keep_alive = True
         self.watcher_thread = threading.Thread(target=self._watch_serial_connection, daemon=True)
         self.watcher_thread.start()
-
+
```
Comment (ctaqvist) [inline Embedded/RaspberryPI/communication/serial_helper.py:18]:
```suggestion

```

File: Embedded/RaspberryPI/commands/rover/power.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        # Skicka till app först
+        websocket.send(json.dumps(data))
+         # Stäng av Raspberry Pi
+        os.system("sudo shutdown now")   
+    else:
+        print("[SleepMode] Battery ok - not sleeping.")
+        return websocket.send(json.dumps(data))
+
+def handle_low_battery_warning(trigger: bool, websocket):
+
```
Comment (ctaqvist) [inline Embedded/RaspberryPI/commands/rover/power.py:34]:
```suggestion
```

File: Embedded/RaspberryPI/commands/rover/power.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    #Om batterinivå är över 20% skickas en signal till appen att Pi inte ska stängas av.
+    data = {
+        "type": "battery_level",
+        "value": level
+    }
+    print(f"[Battery] Level received: {level}")
+    return websocket.send(json.dumps(data))
+
+def handle_sleep_signal(trigger: bool, websocket):
+
```
Comment (ctaqvist) [inline Embedded/RaspberryPI/commands/rover/power.py:17]:
```suggestion
```

File: Embedded/RaspberryPI/commands/rover/power.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
+import os
+import json
+
+def handle_battery_level(level: float, websocket):
+
```
Comment (ctaqvist) [inline Embedded/RaspberryPI/commands/rover/power.py:5]:
```suggestion
```

File: Embedded/RaspberryPI/commands/rover/handler.py [hunk 5]
```diff
@@ -2,11 +2,18 @@
 from ..camera.stream import *
 from .lights import *
 from .forwarder import *
+from .battery import handle_incoming_sleep_command, send_battery_level, send_sleep_mode, send_low_battery_warning
 
 async def process_command(websocket, command, params):
```
Comment (ctaqvist) [inline Embedded/RaspberryPI/commands/rover/handler.py:7]: needs rebasing with master

**Review by George3235 — CHANGES_REQUESTED** [review]
Det här är fel tänkt. Det är Arduinon som skickar battery level till Pi:n och inte tvärtom

**Comment by ctaqvist** [comment]
> Det här är fel tänkt. Det är Arduinon som skickar battery level till Pi:n och inte tvärtom

För att förtydliga lite mer. 

Servern kan inte skicka batteri varningar till raspberry pin. det ska den inte behöva hantera, Arduinon skickar kontinuerligt batteri nivåer, hur ofta, det avgör ni, till RPIn som sedan kollar om batteri nivån är för låg och tar steg för att hantera det då. Batteri nivån bör alltid skickas till servern vid uppdatering av nivån från arduinon skickat till RPin. @George3235 kan du bekräfta / lägga till mer info för att förtydliga kanske

**Comment by malinradstrom** [comment]
Closing this PR because the branch will not be used.  See branch 263/auto shutdown low battery
